File: packages/java/core/src/test/java/dev/configdiscovery/ConformanceTest.java
package dev.configdiscovery;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.File;
import java.nio.file.Path;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.Test;

/**
 * The cross-language conformance suite, run as part of this module's own tests.
 *
 * <p>A conformance regression should fail {@code ./gradlew test}, not only CI.
 */
class ConformanceTest {

    @Test
    void everyFixturePasses() throws Exception {
        Assumptions.assumeTrue(onPath("node"), "the conformance runner needs node");

        // Gradle runs tests with the module directory as the working directory, so walk up
        // until the runner is in sight rather than counting levels.
        Path repoRoot = Path.of("").toAbsolutePath();
        while (repoRoot != null && !java.nio.file.Files.exists(repoRoot.resolve("packages/spec/runner/run.mjs"))) {
            repoRoot = repoRoot.getParent();
        }
        Assumptions.assumeTrue(repoRoot != null, "could not locate the repository root");
        Process process = new ProcessBuilder(
                        "node", repoRoot.resolve("packages/spec/runner/run.mjs").toString(), "--probe", "java")
                .directory(repoRoot.toFile())
                .redirectErrorStream(true)
                .start();
        String output = new String(process.getInputStream().readAllBytes());
        assertEquals(0, process.waitFor(), output);
    }

    private static boolean onPath(String command) {
        String path = System.getenv("PATH");
        if (path == null) {
            return false;
        }
        for (String directory : path.split(File.pathSeparator)) {
            if (new File(directory, command).canExecute()) {
                return true;
            }
        }
        return false;
    }
}

File: packages/golang/conformance_test.go
package configdiscovery

import (
	"os/exec"
	"path/filepath"
	"testing"
)

// The cross-language conformance suite, run as part of this package's own tests. A conformance
// regression should fail `go test`, not only CI.
func TestConformanceSuite(t *testing.T) {
	if _, err := exec.LookPath("node"); err != nil {
		t.Skip("the conformance runner needs node")
	}
	repoRoot, err := filepath.Abs("../..")
	if err != nil {
		t.Fatal(err)
	}
	cmd := exec.Command("node", filepath.Join(repoRoot, "packages/spec/runner/run.mjs"), "--probe", "golang")
	cmd.Dir = repoRoot
	output, err := cmd.CombinedOutput()
	if err != nil {
		t.Fatalf("conformance suite failed: %v\n%s", err, output)
	}
}
